README.md
lds-scriptures
==============

A Chrome Extension that allows you to download and view scripture content from lds.org

I originally planned to put this on the Chrome Web Store, but I discovered that they are now blocking many of the API features I need for the app to operate. The source code is provided for those that would like to download and install it into their own Chrome installs, circumventing the Web Store.

I no longer plan to develop the extension anymore, so I also hope that it will be forked from here and improved. If you would like to submit pull requests from me, go ahead. I would love for the project to live on.

#Installation instructions
Check back later.


xml/changelog.txt
Mormon Documentation Project - LDS Scriptures
---------------------------------------------

2.5.0
-----

- First custom XML release.

- Added Microsoft Excel 97/2000/XP (.xls) and OpenDocument Spreadsheet
(.ods) formats to the export list

- PostgreSQL:
	* used smallint instead of int where possible
	* encoding set to LATIN1
	* cleaner database dump
	* changed all fields to NOT NULL

- MySQL:
	* Removed all NULL values

- SQLite:
	* All NULL values changed to blank

- CSV:
	* Removed the newlines in the CSV files on lds_scriptures_books,
	which may have caused some problems when someone is trying to
	split up the rows based on \n or \r.

- Added 'pilcrow' field to lds_scriptures_verses: Now instead of
verse_scripture having &#142;, the exports have a boolean flag so whoever
uses the data can decide if they want to display the pilcrow or not.
Affected 2,964 verses (all in the Old and New Testaments).

- Text exports (CSV, XML, TeX, ODS, XLS) now have files for both the
relational tables AND a non-relational file (lds_scriptures).  Basically,
it means that all the data from all the tables is in one file in addition
to the regular export of each table individually.  The old releases
would usually have one or the other.  If you're interested in creating
a non-relational table for the databases, see the 'Notes' section on
the website.  Exporting an additional table would increase the export
size greatly, so they are not included in the database downloads.

- Fixed verses where two spaces were between two unpunctuated words.
See the section 'Errata' in the documentation for details about the
punctuation and spacing in the scriptures.

Affected 75 verses:

Num. 6:20, Deut. 5:15, Deut. 31:17, Judg. 16:17, Isa. 37:21, Jer. 34:12,
Ezek. 37:14, 1 Ne. 4:20, 1 Ne. 4:29, 1 Ne. 8:21, 1 Ne. 12:18, 1 Ne. 16:23,
1 Ne. 17:52, 1 Ne. 21:12, 2 Ne. 10:19, 2 Ne. 20:29, 2 Ne. 28:12, Jacob
4:9, Mosiah 2:15, Mosiah 2:32, Mosiah 2:37, Mosiah 4:9, Mosiah 4:14,
Mosiah 7:21, Mosiah 20:13, Mosiah 21:26, Mosiah 25:13, Mosiah 26:12,
Alma 2:12, Alma 2:24, Alma 5:9, Alma 5:57, Alma 15:18, Alma 16:9, Alma
17:27, Alma 18:16, Alma 18:38, Alma 19:16, Alma 26:12, Alma 30:28, Alma
30:33, Alma 34:27, Alma 34:33, Alma 39:11, Alma 47:10, Alma 47:34, Alma
50:11, Alma 50:23, Alma 50:29, Alma 51:31, Alma 57:17, Alma 62:51, Alma
63:12, Hel. 1:21, Hel. 1:32, Hel. 3:30, Hel. 3:36, Hel. 5:41, Hel. 8:11,
Hel. 13:21, 3 Ne. 1:15, 3 Ne. 6:30, 3 Ne. 7:12, 3 Ne. 8:14, 3 Ne. 9:10,
3 Ne. 11:38, 3 Ne. 13:16, 3 Ne. 17:18, 3 Ne. 19:20, 3 Ne. 23:13, 3
Ne. 26:17, Morm. 9:36, Ether 3:16, Ether 8:11, Ether 9:31

- Fixed verses that still had verse numbers in the beginning of the text.

Affected 133 verses:

Enos 1:14, D&C 76:100, D&C 76:101, D&C 76:102, D&C 76:103, D&C 76:104,
D&C 76:105, D&C 76:106, D&C 76:107, D&C 76:108, D&C 76:109, D&C 76:110,
D&C 76:111, D&C 76:112, D&C 76:113, D&C 76:114, D&C 76:115, D&C 76:116,
D&C 76:117, D&C 76:118, D&C 76:119, D&C 84:100, D&C 84:101, D&C 84:102,
D&C 84:103, D&C 84:104, D&C 84:105, D&C 84:106, D&C 84:107, D&C 84:108,
D&C 84:109, D&C 84:110, D&C 84:111, D&C 84:112, D&C 84:113, D&C 84:114,
D&C 84:115, D&C 84:116, D&C 84:117, D&C 84:118, D&C 84:119, D&C 84:120,
D&C 88:100, D&C 88:101, D&C 88:102, D&C 88:103, D&C 88:104, D&C 88:105,
D&C 88:106, D&C 88:107, D&C 88:108, D&C 88:109, D&C 88:110, D&C 88:111,
D&C 88:112, D&C 88:113, D&C 88:114, D&C 88:115, D&C 88:116, D&C 88:117,
D&C 88:118, D&C 88:119, D&C 88:120, D&C 88:121, D&C 88:122, D&C 88:123,
D&C 88:124, D&C 88:125, D&C 88:126, D&C 88:127, D&C 88:128, D&C 88:129,
D&C 88:130, D&C 88:131, D&C 88:132, D&C 88:133, D&C 88:134, D&C 88:135,
D&C 88:136, D&C 88:137, D&C 88:138, D&C 88:139, D&C 88:140, D&C 88:141,
D&C 101:100, D&C 101:101, D&C 107:100, D&C 124:100, D&C 124:101, D&C
124:102, D&C 124:103, D&C 124:104, D&C 124:105, D&C 124:106, D&C 124:107,
D&C 124:108, D&C 124:109, D&C 124:110, D&C 124:111, D&C 124:112, D&C
124:113, D&C 124:114, D&C 124:115, D&C 124:116, D&C 124:117, D&C 124:118,
D&C 124:119, D&C 124:120, D&C 124:121, D&C 124:122, D&C 124:123, D&C
124:124, D&C 124:125, D&C 124:126, D&C 124:127, D&C 124:128, D&C 124:129,
D&C 124:130, D&C 124:131, D&C 124:132, D&C 124:133, D&C 124:134, D&C
124:135, D&C 124:136, D&C 124:137, D&C 124:138, D&C 124:139, D&C 124:140,
D&C 124:141, D&C 124:142, D&C 124:143, D&C 124:144, D&C 124:145

- Cleaned up the documentation.

- Fixed the newlines on the documents in the .zip exports.

2.4.0
-----
- Fixed the files in .zip packages so they use DOS file formats.
The documents and exports should be viewable in Notepad now.

- Fixed the broken XML .zip export (v2.3.0 wouldn't unzip)

- Added lds_org to lds_scriptures_volumes and lds_scriptures_books,
stores the naming scheme for the directories used on the Church's
Scriptures website.  See http://scriptures.lds.org/

- Added num_books, num_chapters, num_verses to lds_scriptures_volumes

- Added num_chapters, num_verses to lds_scriptures_books

- Cleaned up the installation documentation for MySQL + Windows

2.3.0
-----
- New standardized book and verse titles, based on Church's "Abbreviations
and Designations in Footnotes and Index".  The new standardization fixes
a large number of questions as to how the books should be referenced in
book_title and book_title_short, what the verse_title value for the Four
Gospels would be (St John, St. John, Saint John, John), and fixes old
values in lds_scriptures_books (ex: 1 Chronicles).  Thanks to everyone
who made me aware of the issues. :)

- MySQL: Added a FULLTEXT index on lds_scriptures_verses.verse_scripture.

See the MySQL documentation on how this can really help when writing
functions to search the database.
http://dev.mysql.com/doc/mysql/en/Fulltext_Search.html

- Fixed bug where 62 verses from New Testament were missing.

- Fixed bug where verse_title and verse_title_long were incorrect in
lds_scriptures_verses

- Much needed documenation finally created in DocBook format.
http://mdp.nephi.org/scriptures/docs/


src/js/ref.js
//This document is at the bottom of the page so it loads after the
// DOM loads. It is NOT in charge of loading the HTML from the
// background page, but is responsible for the keyboard shortcuts,
// styles, toolbar, and any other operations on the ref.html page.
//
// Is only to be loaded at the bottom of /ref.html

function loadpage()
{
	read();
	searchMode_leave();
	search = document.getElementById('search');
	search.addEventListener('focus', searchMode_enter, false);
	search.addEventListener('blur', searchMode_leave, false);
}

if (search != '')
	tabBookmarkCheck(search);
verseSearch = 0;
if (GetStyleMode('CSSFile'))
	document.head.appendChild(GetStyleTag('CSSFile'));
if (GetStyleMode('CustomCSS'))
	document.head.appendChild(GetStyleTag('CustomCSS'));
if (GetStyleMode('Simple'))
	document.head.appendChild(GetStyleTag('Simple'));
	
function keypress(e)
{
	if (e.keyCode == 110) //n
	{
		ScrollToVerse(scrolledVerse + 1);//Next verse
		e.preventDefault();
	}
	if (e.keyCode == 98) //b
	{
		ScrollToVerse(scrolledVerse - 1);//Previous verse
		e.preventDefault();
	}
	if (e.keyCode == 122) //z
	{
		location.href = document.getElementById('nextlink').href;//Next chapter
		e.preventDefault();
	}
	if (e.keyCode == 97) //a
	{
		location.href = document.getElementById('backlink').href;//Previous chapter
		e.preventDefault();
	}
	if (e.keyCode == 113) //q
	{
		location.href = document.getElementById('uplink').href;//Up a level
		e.preventDefault();
	}
	if (e.keyCode == 117) //u
	{
		updatebookmark();
		e.preventDefault();
	}
	if (e.keyCode == 115) //s
	{
		document.getElementById('search').focus();//Focus the search bar
		e.preventDefault();
	}
	if (e.keyCode == 120) //x
	{
		var a = document.getElementById('autoscroll')
		a.checked = !a.checked;
		togglescroll(a.checked);
		e.preventDefault();
	}
	console.debug('Key Pressed: ' + e.keyCode);
}

scrolledVerse = 0;

function ScrollToVerse(num)
{
	var theElement = document.getElementById(String(num));
	var selectedPosY = 0;

	while(theElement != null)
	{
		selectedPosY += theElement.offsetTop;
		theElement = theElement.offsetParent;
	}
	window.scrollTo(selectedPosX,selectedPosY);
}
function window_scroll()
{
	var vc = document.getElementById('0').children;
	for (var x = 0; x < vc.length; x++)
	{
		var el = document.getElementById(String(x));
		if (el == undefined || el == null)
			continue;
		var selectedPosY = 0;
		while (el != null)
		{
			selectedPosY += el.offsetTop;
			el = el.offsetParent;
		}
		if (selectedPosY > window.pageYOffset + 10)
		{
			scrolledVerse = x - 1;
			if (scrolledVerse < 1)
				scrolledVerse = 0
			//console.debug('Scrolled To: ' + scrolledVerse);
			break;
		}
	}
	
	// This section sets the hash code, but prevents it from scrolling to that verse
	var el = document.getElementById(String(scrolledVerse));
	if (el)
	{
		el.id = '';
		location.hash = scrolledVerse;
		el.id = String(scrolledVerse);
	}
	
	// If the page is not being automatically scrolled, then 
	// set the desired position to where it is at.
	if (!scrolling)
		desiredScroll = window.pageYOffset;
}

chrome.runtime.onMessage.addListener(function(request) {
	if (request.request == 'update style')
	{
		//Must update the styles

		//Should remove any style tags, but the default one.
		var toremove = document.getElementsByTagName('link')[1];
		while (toremove)
		{
			document.head.removeChild(toremove);
			toremove = document.getElementsByTagName('link')[1];
		}

		toremove = document.getElementsByTagName('style')[0];
		while (toremove)
		{
			document.head.removeChild(toremove)
			toremove = document.getElementsByTagName('style')[0];
		}
		
		//Add whichever style tags are allowed by the settings
		if (GetStyleMode('CSSFile'))
			document.head.appendChild(GetStyleTag('CSSFile'));
		if (GetStyleMode('CustomCSS'))
			document.head.appendChild(GetStyleTag('CustomCSS'));
		if (GetStyleMode('Simple'))
			document.head.appendChild(GetStyleTag('Simple'));
	}
});

function searchMode_enter()
{
	document.removeEventListener('keypress', keypress, false);
 
	//Add a specific class for specifying that the bar is shown
	document.getElementsByClassName('navbar')[0].classList.add('open')
}
function searchMode_leave()
{
	document.addEventListener('keypress', keypress, false);

	//Remove the specific class for specifying that the bar is shown
	document.getElementsByClassName('navbar')[0].classList.remove('open')
}

var desiredScroll = 0;
var scrolling = false;
var scrollSpeed = 1;//This is the speed of the auto scroll
					//TODO make shortcut keys and a control on the panel to control the speed

function togglescroll(scroll)
{
	if (scroll)
		scrollTimer = setInterval(scrollTimer, 33);
	else clearInterval(scrollTimer);
	scrolling = scroll;
}
function scrollTimer()
{
	desiredScroll += scrollSpeed;
	window.scrollTo(window.pageXOffset, desiredScroll);
}

loadpage();
